BMI model init config pydantic models

aaraney opened the pull request:
This PR adds pydantic models for currently supported NGEN BMI models (i.e. `CFE`). This work makes it possible to serialize, deserialize, validate, and programmatically access supported BMI model configuration files in a consistent manor.

This work relies on #43. So, it should not be merged until #43 has.

~I still have a few models to port over, so for now, I am going to leave this as a draft and gradually port models to this PR.~

## Additions

- pydantic model of [`CFE`](https://github.com/NOAA-OWP/cfe/)'s configuration file.
- "" [`PET`](https://github.com/NOAA-OWP/evapotranspiration/)'s configuration file.
- "" [`NoahOWP`](https://github.com/NOAA-OWP/noah-owp-modular/tree/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c)'s configuration file.
- "" [`LSTM`](https://github.com/NOAA-OWP/lstm)'s configuration file.


## Todos

- CFE has rootzone-based actual evapotranspiration (AET) and ice-fraction based runoff options that were not super clear. I have marked TODOs to go back and add these options once I have a better grasp on these options. I dont think these TODOs should hold up this PR.
- modify LSTM config model once a better understanding of how LSTM is to be configured in practice. Based on the configs present in https://github.com/NOAA-OWP/lstm it is unclear when certain keys are required. It is safe to merge this and leave this for later if so we choose.
- add LSTM tests

## Checklist

- [x] PR has an informative and human-readable title
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code can be automatically merged (no conflicts)
- [ ] Code follows project standards (link if applicable)
- [x] Passes all existing automated tests
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] Project documentation has been updated (including the "Unreleased" section of the CHANGELOG)
- [x] Reviewers requested with the [Reviewers tool](https://help.github.com/articles/requesting-a-pull-request-review/) :arrow_right:

aaraney commented on python/ngen_conf/src/ngen/config/init_config/noahowp_options.py:
@SnowHydrology, when you get a change, could you take a look at the enumeration names i've gone with here and provide any feedback / suggestions? In short, these enums are more or less human friendly and _hopefully_ meaningful mappings to associated Noah OWP Modular `model_options`. They will still show up in the config file as integer values, but users who are playing around with a configuration file programmatically might appreciate meaningful names. Thanks in advance!

hellkite500 commented on python/ngen_conf/src/ngen/config/init_config/lstm.py:
```diff
@@ -0,0 +1,123 @@
+    # NOTE: listed as post-processing / plotting only
+    # lat: float
+    # lon: float
+    area_sqkm: float
```
My understanding of the LSTM is that depending on which `train_cfg_file` is used, the required parameters can be different...

aaraney replied:
That is my understand as well. I don't know all the possible combinations of parameters, so if we want to leave `LSTM` out of this PR and add it at a later date, im fine with that. Thoughts?

jmframe replied:
I don't know the context here, but I would like to. In any rate, why is LSTM configuration files needed in a calibration repository anyway? LSTM would be trained via back propogation, no calibration necessary. In terms of the required parameters, it pretty much comes down to what would be available. It became clear that the majority of the CAMELS attributes would not be available in the hydrofabric, so we trained on only a couple attributes we knew would be available (slope and elevation, etc.). However, it appears there are several NOAH-MP parameters that are available in the hydrofabric, so I think the best thing to do would be to train an LSTM on those. Which I am planning to do soon-ish.

hellkite500 replied:
because this repo no longer just does calibration things, there is a configuration package developed in this repo that is used by the calibration, but also by other workflows that need to generate configuration files and components programatically.

aaraney replied:
Welll @jmframe , the repo name `ngen-cal` is not the most descriptive. This repo has become a catchall for `ngen` configuration _and_ calibration related concerns. The goal of the `ngen.init_config` package is to capture the schema for `bmi` module configuration files (e.g. `LSTM`). This is desirable so we can fully validate an `ngen` realization config and all linked `bmi` config files.

With that in mind, it would be extremely helpful to know _which_ parameters _can_ appear in a config file and _when_ (i.e. if that is conditional).

jmframe replied:
Okay, I see. That makes sense. I think that a re-visit to the LSTM module with the maturation of NGen and Hydrofabric over the past two years is warranted. It would be great to have a chat with @mikejohnson51 and get an update on what static catchment attributes are always going to be available. Then re-train a few models and do some experimenting with this in mind. Then make a decision on the best keys to include. But for now, the LSTM configuration should have at least: lat, lon, area_sqkm, elev_mean, slope_mean

aaraney replied:
Thanks for the input, @jmframe! For now, I will only include the keys you mentioned and in the future if things change, we can add others.

hellkite500 replied:
Let's open an issue to come back to LSTM configs and see if we can use a target `train_cfg_file` to determine fields required to support that particular model.  Having a "default" for now is a good place to start.

hellkite500 commented on python/ngen_conf/src/ngen/config/init_config/noahowp.py:
```diff
@@ -0,0 +1,248 @@
+    # source: https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/NamelistRead.f90#L36
+    croptype: int = 0
+    sfctyp: LandSurfaceType  #    = 1
+    soilcolor: int  # = 4
```
soil color...interesting...

aaraney replied:
Yeah???

hellkite500 replied:
I double checked the source and it surely is `soilcolor`.
Per @SnowHydrology 's explanation, it "provide different values of wet vs. dry albedo based on the soil color number"

hellkite500 commented on python/ngen_conf/src/ngen/config/init_config/noahowp_options.py:
```diff
@@ -0,0 +1,282 @@
+    topmodel_with_equilibrium_water_table = 2
+    original_surface_and_subsurface_runoff = 3
+    bats_surface_and_subsurface_runoff = 4
+    Miguez_Macho_Fan_groundwater_scheme = 5
```
can we use lowercase in these enums to be consistent?

aaraney replied:
For sure. Changing that now.

hellkite500 commented on python/ngen_conf/src/ngen/config/init_config/utils.py:
```diff
@@ -0,0 +1,100 @@
+# 4.05[] -> group 0: 4.05 group 2: ""
+# 2e5[m] -> group 0: 2 group 1 e5 group 2: m
+# 1.8e-05[m h-1] -> group 0: 1.8 group 1: e-05 group 2: m h-1
+_FLOAT_UNIT_PAIR_RE_PATTERN = r"([-]?\d*\.\d+|[-]?\d+)([Ee][-+]?\d+)?\[([^\]]*)\]"
```
checked this out on https://regex101.com/ and it looks pretty solid

aaraney replied:
Thanks for checking it!

hellkite500 reviewed (approved):
In general this looks useable as is, but may need to consider refactoring the LSTM config at some point

aaraney commented:
@hellkite500, I think I addressed everything that you asked for. I am meeting with @SnowHydrology tomorrow to talk about other changes that might need to be made in this PR. After that I think we should be about ready to merge.

hellkite500 commented:
Looks like this will also need a rebase to resolve the setup.cfg conflict.

aaraney commented:
> Looks like this will also need a rebase to resolve the setup.cfg conflict.

@hellkite500, done.

aaraney commented:
The `3.7` tests are failing because `ngen.init_config` requires `>=3.8`. I need to look back and see why that is the case and if the version requirement can be lowered to `>=3.7`.

aaraney commented:
@hellkite500, okay the tests are passing now and 3.7 support has been remedied via #58. I think we should be good to go now! Thanks again!

SnowHydrology reviewed (commented):
Looks mostly good to me. All tests pass but I added a few line-specific comments. Could also benefit from documentation and build instructions.

@ajkhattak, can you double-check the `cfe.py` file?

ajkhattak commented:
> @ajkhattak, can you double-check the `cfe.py` file?

I am not sure why this is commented out, if we are not explicitly running any PET module then CFE takes PET (i.e., `EVAPOTRANS`) from NOAH_OM as an input or am I missing something here
`#"water_potential_evaporation_flux": "EVAPOTRANS",`

One other comment I have is on the sloth.py
`model_params: Optional[Mapping[str, str]]`
Should the last str in the Mapping list be float? or Pydantic will convert float to str or vice versa. 

I am not very familiar with Pydantic, so not 100% sure about the format conversion, etc.

ajkhattak commented on python/ngen_conf/src/ngen/config/init_config/cfe.py:
```diff
@@ -0,0 +1,165 @@
+    # saturated soil moisture content
+    soil_params_smcmax: FloatUnitPair[m_per_m]
+    # wilting point soil moisture content
+    soil_params_wltsmc: FloatUnitPair[m_per_m]
```
the parameter `refkdt` is missing here (optional parameter), the default value is 3.0 set inside cfe.c

ajkhattak commented on python/ngen_conf/src/ngen/config/init_config/cfe.py:
```diff
@@ -0,0 +1,165 @@
+
+    # TODO: add support for below fields
+
+    # # when `CFE coupled to SoilMoistureProfile`
+    # aet_rootzone: bool  # True, true, 1
+    # # layer of the soil that is the maximum root zone depth. That is, the depth of the layer where the AET is drawn from
+    # max_root_zone_layer: FloatUnitPair[m]
+    # # an array of depths from the surface. Example, soil_layer_depths=0.1,0.4,1.0,2.0
+    # soil_layer_depths: List[float]
+    # # `ice-fraction based runoff` | when `CFE coupled to SoilFreezeThaw`
+    # sft_coupled: bool  # True, true, 1
```
@aaraney I know you mentioned you plan to do these, they will be needed when cfe is coupled to SoilMoistureProfiles and SoilFreezeThaw models otherwise not.

aaraney replied:
@ajkhattak, I assume that the presence of these fields in a configuration file indicates to the model that `cfe` is couple to `SoilMoistureProfiles`? Or is there a flag in the config file I can use to determine if these fields should be included or not?

ajkhattak replied:
You are right.
1. If `aet_rootzone=true` is given in the config file, then `CFE` is coupled to `SoilMoistureProfiles` and two additional parameters need to be provided `soil_layer_depths` (1D array) and `max_root_zone_layer` (double).
2. If `sft_coupled=true` is provided in the config file, then CFE is coupled to `SoilFreezeThaw` model, now if the `surface_partitioning_scheme=Schaake`, then one additional parameter `ice_content_threshold` is needed in the config file. If `surface_partitioning_scheme=Xinanjiang` then no additional parameters are needed.

Let me know if you have any questions...

ajkhattak commented on python/ngen_conf/src/ngen/config/init_config/cfe.py:
```diff
@@ -0,0 +1,165 @@
+class CFEXinanjiang(CFEBase):
+    # direct runoff
+    surface_partitioning_scheme: Literal["Xinanjiang"] = "Xinanjiang"
+    a_xinanjiang_inflection_point_parameter: int
+    b_xinanjiang_shape_parameter: int
+    x_xinanjiang_shape_parameter: int
```
should the type be `float` here (or doesn't matter)? in cfe these parameters are double

aaraney replied:
Great catch, yes they should be `float`s.

aaraney commented on python/ngen_conf/src/ngen/config/init_config/noahowp_options.py:
```diff
@@ -0,0 +1,282 @@
+
+    noah_mp = 1
+    one_way_coupled_hydrostatic = 2
+    two_way_coupled_hydrostatic_not_implemented = 2
```
@SnowHydrology, I think it makes the most sense to comment it out and leave it for now. Once it has been implemented we can remove the comment and re-release.

SnowHydrology replied:
Sounds good, thanks!

aaraney commented on python/ngen_conf/src/ngen/config/init_config/noahowp.py:
```diff
@@ -0,0 +1,248 @@
+    # crop type (SET TO 0, no crops currently supported)
+    # source: https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/NamelistRead.f90#L36
+    croptype: int = 0
+    sfctyp: LandSurfaceType  #    = 1
```
@SnowHydrology, the [source](https://github.com/NOAA-OWP/noah-owp-modular/blob/30d0f53e8c14acc4ce74018e06ff7c9410ecc13c/src/NamelistRead.f90#L36
) of `noahowp` listed the meaning of the two `sfctyp` variants (soil or lake), but the others did not.

```fortran
  integer            :: croptype           ! crop type (SET TO 0, no crops currently supported)
```

Can point me in the direction of the other fields variant names? It would be great to include them too.

SnowHydrology replied:
Ah, that makes sense. The list of `vegtyp` options depends on `veg_class_name`.

aaraney commented on python/ngen_conf/src/ngen/config/init_config/noahowp.py:
```diff
@@ -0,0 +1,248 @@
+    isltyp: int  #    = 1
+    nsoil: int  #     = 4
+    nsnow: int  #     = 3
+    nveg: int  #      = 20
```
Yeah for sure! modififed IGBP modis noah is 20 and USGS is 27 correct?

SnowHydrology replied:
Correct!

aaraney commented:
> I am not sure why this is commented out, if we are not explicitly running any PET module then CFE takes PET (i.e., `EVAPOTRANS`) from NOAH_OM as an input or am I missing something here `#"water_potential_evaporation_flux": "EVAPOTRANS",`
> 
> One other comment I have is on the sloth.py `model_params: Optional[Mapping[str, str]]` Should the last str in the Mapping list be float? or Pydantic will convert float to str or vice versa.
> 
> I am not very familiar with Pydantic, so not 100% sure about the format conversion, etc.


@ajkhattak, I think both of your questions are unrelated to this PR. Admittedly this is a little confusing given that in this PR I added a second module named `cfe.py`. I believe your question is about this code, right? I moved your question into #61 and pinged @hellkite500 to take a look at that.

https://github.com/NOAA-OWP/ngen-cal/blob/8d46c053b4a6d553d0b6e2c4ac539b5c092a2a63/python/ngen_conf/src/ngen/config/cfe.py#L29

aaraney commented:
@SnowHydrology and @ajkhattak, thanks you both for your careful eyes! Ive made the changes that were requested by both of you. If either of you could give it a quick re-review glance, that would be greatly appreciated!

ajkhattak commented:
@aaraney looks good to me

SnowHydrology commented:
> @aaraney looks good to me

Ditto

aaraney commented:
Thanks, @ajkhattak and @SnowHydrology! Cheers
